recipes/gsoap/all/test_package/conanfile.py

```diff
@@ -1,33 +1,42 @@
 from conan import ConanFile
-from conan.tools.build import can_run, cross_building
+from conan.tools.build import can_run
 from conan.tools.cmake import CMake, cmake_layout
+from conan.tools.env import VirtualBuildEnv, VirtualRunEnv
 import os
 
 
-class TestGsoapConan(ConanFile):
+class TestPackageConan(ConanFile):
     settings = "os", "arch", "compiler", "build_type"
-    generators = "CMakeDeps", "CMakeToolchain", "VirtualRunEnv"
+    generators = "CMakeDeps", "CMakeToolchain"
     test_type = "explicit"
 
     def layout(self):
         cmake_layout(self)
 
     def requirements(self):
-        self.requires(self.tested_reference_str)
+        self.requires(self.tested_reference_str, run=can_run(self))
+
+    def build_requirements(self):
+        if not can_run(self):
+            self.tool_requires(self.tested_reference_str)
+
+    def generate(self):
+        VirtualRunEnv(self).generate()
+        if can_run(self):
+            VirtualRunEnv(self).generate(scope="build")
+        else:
+            VirtualBuildEnv(self).generate()
 
     def build(self):
-        # TODO: Add some test for the cross-building scenario
-
-        if not cross_building(self):
-            calc_wsdl = os.path.join(self.source_folder, "calc.wsdl")
-            self.output.info(f"Generating code from WSDL '{calc_wsdl}'")
-            self.run(f"wsdl2h -o calc.h {calc_wsdl}", env="conanrun")
-            import_dir = os.path.join(self.deps_cpp_info["gsoap"].rootpath, "bin", "import")
-            self.run(f"soapcpp2 -j -CL -I{import_dir} calc.h", env="conanrun")
-
-            cmake = CMake(self)
-            cmake.configure()
-            cmake.build()
+        calc_wsdl = os.path.join(self.source_folder, "calc.wsdl")
+        self.output.info(f"Generating code from WSDL '{calc_wsdl}'")
+        self.run(f"wsdl2h -o calc.h {calc_wsdl}")
+        import_dir = os.path.join(self.dependencies["gsoap"].package_folder, "bin", "import")
```

Comment:
@memsharded @jcar87 Do you know how to migrate this logic to conan v1? self.dependencies is None in build() of test package, but it seems ok for conan v2 client. It's quite problematic.